#### Pandas.apply()

Pandas.apply allow the users to pass a function and apply it on every single value of the Pandas series. It comes as a huge improvement for the pandas library as this function helps to segregate data according to the conditions required due to which it is efficiently used in data science and machine learning

Syntax:

s.apply(func, convert_dtype=True, args=())

Parameters:

    func: .apply takes a function and applies it to all values of pandas series.
    convert_dtype: Convert dtype as per the function’s operation.
    args=(): Additional arguments to pass to function instead of series.
    Return Type: Pandas Series after applied function/operation. 

The following example passes a function and checks the value of each element in series and returns low, normal or High accordingly.

In [1]:

import pandas as pd 
  
# reading csv 
s = pd.read_csv("stock.csv", squeeze = True) 

print(s.head())

# defining function to check price 
def fun(num): 
  
    if num<200: 
        return "Low"
  
    elif num>= 200 and num<400: 
        return "Normal"
  
    else: 
        return "High"

    
# passing function to apply and storing returned series in new 
new = s.apply(fun) 
  
# printing first 3 element 
print(new.head()) 
  
# printing elements somewhere near the middle of series 
print(new[1400], new[1500], new[1600]) 
  
# printing last 3 elements 
print(new.tail(5)) 


0    50.12
1    54.10
2    54.65
3    52.38
4    52.95
Name: Stock Price, dtype: float64
0    Low
1    Low
2    Low
3    Low
4    Low
Name: Stock Price, dtype: object
Normal Normal Normal
3007    High
3008    High
3009    High
3010    High
3011    High
Name: Stock Price, dtype: object


In the following example, a temporary anonymous function is made in .apply itself using lambda. It adds 5 to each value in series and returns a new series.

In [2]:

import pandas as pd 
s = pd.read_csv("stock.csv", squeeze = True) 

# adding 5 to each value 
new = s.apply(lambda num : num + 5) 
  
# printing first 5 elements of old and new series 
print(s.head(), '\n', new.head()) 
  
# printing last 5 elements of old and new series 
print('\n\n', s.tail(), '\n', new.tail()) 


0    50.12
1    54.10
2    54.65
3    52.38
4    52.95
Name: Stock Price, dtype: float64 
 0    55.12
1    59.10
2    59.65
3    57.38
4    57.95
Name: Stock Price, dtype: float64


 3007    772.88
3008    771.07
3009    773.18
3010    771.61
3011    782.22
Name: Stock Price, dtype: float64 
 3007    777.88
3008    776.07
3009    778.18
3010    776.61
3011    787.22
Name: Stock Price, dtype: float64


#### Apply function to every row in a Pandas DataFrame

In [4]:

# Import pandas package  
import pandas as pd 
  
# Function to add 
def add(a, b, c): 
    return a + b + c 
  
def main(): 
      
    # create a dictionary with 
    # three fields each 
    data = { 
            'A':[1, 2, 3],  
            'B':[4, 5, 6],  
            'C':[7, 8, 9] } 
      
    # Convert the dictionary into DataFrame  
    df = pd.DataFrame(data) 
    print("Original DataFrame:\n", df) 
      
    df['add'] = df.apply(lambda row : add(row['A'], 
                     row['B'], row['C']), axis = 1) 
   
    print('\nAfter Applying Function: ') 
    # printing the new dataframe 
    print(df) 

if __name__ == '__main__': 
    main() 


Original DataFrame:
    A  B  C
0  1  4  7
1  2  5  8
2  3  6  9

After Applying Function: 
   A  B  C  add
0  1  4  7   12
1  2  5  8   15
2  3  6  9   18


You can use the numpy function as the parameters to the dataframe as well.

In [5]:

import pandas as pd 
import numpy as np 
   
def main(): 
      
    # create a dictionary with 
    # five fields each 
    data = { 
            'A':[1, 2, 3],  
            'B':[4, 5, 6],  
            'C':[7, 8, 9] } 
      
    # Convert the dictionary into DataFrame  
    df = pd.DataFrame(data) 
    print("Original DataFrame:\n", df) 
      
    # applying function to each row in the dataframe 
    # and storing result in a new coloumn 
    df['add'] = df.apply(np.sum, axis = 1) 
   
    print('\nAfter Applying Function: ') 
    # printing the new dataframe 
    print(df) 
   
if __name__ == '__main__': 
    main() 


Original DataFrame:
    A  B  C
0  1  4  7
1  2  5  8
2  3  6  9

After Applying Function: 
   A  B  C  add
0  1  4  7   12
1  2  5  8   15
2  3  6  9   18


Normalising Data 

In [10]:

# Import pandas package  
import pandas as pd 
  
def normalize(x, y): 
    x_new = ((x - np.mean([x, y])) /
             (max(x, y) - min(x, y))) 
      
    # print(x_new) 
    return x_new 
  
def main(): 
      
    # create a dictionary with three fields each 
    data = { 
        'X':[1, 2, 3],  
        'Y':[45, 65, 89] } 
      
    # Convert the dictionary into DataFrame  
    df = pd.DataFrame(data) 
    print("Original DataFrame:\n", df) 
      
    df['X'] = df.apply(lambda row : normalize(row['X'], 
                                  row['Y']), axis = 1) 
   
    print('\nNormalized:') 
    print(df) 
   
if __name__ == '__main__': 
    main() 


Original DataFrame:
    X   Y
0  1  45
1  2  65
2  3  89

Normalized:
     X   Y
0 -0.5  45
1 -0.5  65
2 -0.5  89


Generate range

In [11]:

import pandas as pd 
import numpy as np 
   
pd.options.mode.chained_assignment = None
  
# Function to generate range 
def generate_range(n): 
      
    # printing the range for eg: 
    # input is 67 output is 60-70 
    n = int(n) 
      
    lower_limit = n//10 * 10
    upper_limit = lower_limit + 10
      
    return str(str(lower_limit) + '-' + str(upper_limit)) 
       
def replace(row): 
    for i, item in enumerate(row): 
          
        # updating the value of the row 
        row[i] = generate_range(item) 
    return row 
           

    
def main(): 
    # create a dictionary with 
    # three fields each 
    data = { 
            'A':[0, 2, 3],  
            'B':[4, 15, 6],  
            'C':[47, 8, 19] } 
      
    # Convert the dictionary into DataFrame  
    df = pd.DataFrame(data) 
   
    print('Before applying function: ') 
    print(df) 
       
    # applying function to each row in 
    # dataframe and storing result in a new coloumn 
    df = df.apply(lambda row : replace(row)) 
       
   
    print('After Applying Function: ') 
    # printing the new dataframe 
    print(df) 

    
if __name__ == '__main__': 
    main() 


Before applying function: 
   A   B   C
0  0   4  47
1  2  15   8
2  3   6  19
After Applying Function: 
      A      B      C
0  0-10   0-10  40-50
1  0-10  10-20   0-10
2  0-10   0-10  10-20


#### Pandas Series.apply()

    Syntax: Series.apply(func, convert_dtype=True, args=(), **kwds)


    Parameter :
    func : Python function or NumPy ufunc to apply.
    convert_dtype : Try to find better dtype for elementwise function results.
    args : Positional arguments passed to func after the series value.
    **kwds : Additional keyword arguments passed to func.

    Returns : Series


Use Series.apply() function to change the city name to ‘Montreal’ if the city is ‘Rio’.

In [12]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the Series 
sr = pd.Series(['New York', 'Chicago', 'Toronto', 'Lisbon', 'Rio']) 
  
# Create the Index 
index_ = ['City 1', 'City 2', 'City 3', 'City 4', 'City 5']  
  
# set the index 
sr.index = index_ 
  
# Print the series 
print(sr) 


City 1    New York
City 2     Chicago
City 3     Toronto
City 4      Lisbon
City 5         Rio
dtype: object


In [13]:

# change 'Rio' to 'Montreal' 
# we have used a lambda function 
result = sr.apply(lambda x : 'Montreal' if x =='Rio' else x ) 
  
# Print the result 
print(result) 


City 1    New York
City 2     Chicago
City 3     Toronto
City 4      Lisbon
City 5    Montreal
dtype: object


 Use Series.apply() function to return True if the value in the given series object is greater than 30 else return False

In [14]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the Series 
sr = pd.Series([11, 21, 8, 18, 65, 18, 32, 10, 5, 32, None]) 
  
# Create the Index 
# apply yearly frequency 
index_ = pd.date_range('2010-10-09 08:45', periods = 11, freq ='Y') 
  
# set the index 
sr.index = index_ 
  
# Print the series 
print(sr) 


2010-12-31 08:45:00    11.0
2011-12-31 08:45:00    21.0
2012-12-31 08:45:00     8.0
2013-12-31 08:45:00    18.0
2014-12-31 08:45:00    65.0
2015-12-31 08:45:00    18.0
2016-12-31 08:45:00    32.0
2017-12-31 08:45:00    10.0
2018-12-31 08:45:00     5.0
2019-12-31 08:45:00    32.0
2020-12-31 08:45:00     NaN
Freq: A-DEC, dtype: float64


In [15]:

# return True if greater than 30 
# else return False 
result = sr.apply(lambda x : True if x>30 else False) 
  
# Print the result 
print(result) 


2010-12-31 08:45:00    False
2011-12-31 08:45:00    False
2012-12-31 08:45:00    False
2013-12-31 08:45:00    False
2014-12-31 08:45:00     True
2015-12-31 08:45:00    False
2016-12-31 08:45:00     True
2017-12-31 08:45:00    False
2018-12-31 08:45:00    False
2019-12-31 08:45:00     True
2020-12-31 08:45:00    False
Freq: A-DEC, dtype: bool


#### Pandas dataframe.aggregate()

Dataframe.aggregate() function is used to apply some aggregation across one or more column. Aggregate using callable, string, dict, or list of string/callables. Most frequently used aggregations are:

    sum: Return the sum of the values for the requested axis
    min: Return the minimum of the values for the requested axis
    max: Return the maximum of the values for the requested axis

    Syntax: DataFrame.aggregate(func, axis=0, *args, **kwargs)

    Parameters:
    func : callable, string, dictionary, or list of string/callables. Function to use for aggregating the data. If a function, must either work when passed a DataFrame or when passed to DataFrame.apply. For a DataFrame, can pass a dict, if the keys are DataFrame column names.
    axis : (default 0) {0 or ‘index’, 1 or ‘columns’} 0 or ‘index’: apply function to each column. 1 or ‘columns’: apply function to each row.

    Returns: Aggregated DataFrame

#### Aggregate ‘sum’ and ‘min’ function across all the columns in data frame

In [16]:

# importing pandas package 
import pandas as pd 
  
# making data frame from csv file 
df = pd.read_csv("nba.csv") 
  
# printing the first 10 rows of the dataframe 
df[:10] 


Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [17]:

# Applying aggregation across all the columns  
# sum and min will be found for each  
# numeric type column in df dataframe 
  
df.aggregate(['sum', 'min']) 


Number      Age    Weight        Salary
sum  8079.0  12311.0  101236.0  2.159837e+09
min     0.0     19.0     161.0  3.088800e+04

In [18]:
df.aggregate(['sum', 'max']) 

Number      Age    Weight        Salary
sum  8079.0  12311.0  101236.0  2.159837e+09
max    99.0     40.0     307.0  2.500000e+07

In Pandas, we can also apply different aggregation functions across different columns. For that, we need to pass a dictionary with key containing the column names and values containing the list of aggregation functions for any specific column

In [19]:

# importing pandas package 
import pandas as pd 
  
# making data frame from csv file 
df = pd.read_csv("nba.csv") 
  
# We are going to find aggregation for these columns 
df.aggregate({"Number":['sum', 'min'], 
              "Age":['max', 'min'], 
              "Weight":['min', 'sum'],  
              "Salary":['sum']}) 


Number   Age    Weight        Salary
max     NaN  40.0       NaN           NaN
min     0.0  19.0     161.0           NaN
sum  8079.0   NaN  101236.0  2.159837e+09

#### Pandas dataframe.mean()

Pandas dataframe.mean() function return the mean of the values for the requested axis. If the method is applied on a pandas series object, then the method returns a scalar value which is the mean value of all the observations in the dataframe. If the method is applied on a pandas dataframe object, then the method returns a pandas series object which contains the mean of the values over the specified axis.

    Syntax: DataFrame.mean(axis=None, skipna=None, level=None, numeric_only=None, **kwargs)


    Parameters :
    axis : {index (0), columns (1)}
    skipna : Exclude NA/null values when computing the result

    level : If the axis is a MultiIndex (hierarchical), count along a particular level, collapsing into a Series

    numeric_only : Include only float, int, boolean columns. If None, will attempt to use everything, then use only numeric data. Not implemented for Series.

    Returns : mean : Series or DataFrame (if level specified) 

Use mean() function to find the mean of all the observations over the index axis

In [21]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the dataframe  
df = pd.DataFrame({"A":[12, 4, 5, 44, 1], 
                   "B":[5, 2, 54, 3, 2],  
                   "C":[20, 16, 7, 3, 8], 
                   "D":[14, 3, 17, 2, 6]}) 
  
# Print the dataframe 
df 


A   B   C   D
0  12   5  20  14
1   4   2  16   3
2   5  54   7  17
3  44   3   3   2
4   1   2   8   6

In [22]:

# Even if we do not specify axis = 0, 
# the method will return the mean over 
# the index axis by default 
df.mean(axis = 0) 


A    13.2
B    13.2
C    10.8
D     8.4
dtype: float64

Use mean() function to find the mean of all the observations over the index axis

In [23]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the dataframe  
df = pd.DataFrame({"A":[12, 4, 5, 44, 1], 
                   "B":[5, 2, 54, 3, 2],  
                   "C":[20, 16, 7, 3, 8], 
                   "D":[14, 3, 17, 2, 6]}) 
  
# Print the dataframe 
df 


A   B   C   D
0  12   5  20  14
1   4   2  16   3
2   5  54   7  17
3  44   3   3   2
4   1   2   8   6

In [24]:

# Even if we do not specify axis = 0, 
# the method will return the mean over 
# the index axis by default 
df.mean(axis = 0) 


A    13.2
B    13.2
C    10.8
D     8.4
dtype: float64

 Use mean() function on a dataframe which has Na values. Also find the mean over the column axis

In [26]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the dataframe  
df = pd.DataFrame({"A":[12, 4, 5, None, 1], 
                   "B":[7, 2, 54, 3, None], 
                   "C":[20, 16, 11, 3, 8], 
                   "D":[14, 3, None, 2, 6]}) 
  
# skip the Na values while finding the mean 
df.mean(axis = 1, skipna = True) 


0    13.250000
1     6.250000
2    23.333333
3     2.666667
4     5.000000
dtype: float64

#### Pandas Series.mean()

Pandas Series.mean() function return the mean of the underlying data in the given Series object.

    Syntax: Series.mean(axis=None, skipna=None, level=None, numeric_only=None, **kwargs)


    Parameter :
    axis : Axis for the function to be applied on.
    skipna : Exclude NA/null values when computing the result.
    level : If the axis is a MultiIndex (hierarchical), count along a particular level, collapsing into a scalar.
    numeric_only : Include only float, int, boolean columns.
    **kwargs : Additional keyword arguments to be passed to the function.

    Returns : mean : scalar or Series (if level specified) 

Use Series.mean() function to find the mean of the underlying data in the given series object.

In [28]:
# importing pandas as pd 
import pandas as pd 
  
# Creating the Series 
sr = pd.Series([10, 25, 3, 25, 24, 6]) 
  
# Create the Index 
index_ = ['Coca Cola', 'Sprite', 'Coke', 'Fanta', 'Dew', 'ThumbsUp'] 
  
# set the index 
sr.index = index_ 
  
# Print the series 
print(sr)

Coca Cola    10
Sprite       25
Coke          3
Fanta        25
Dew          24
ThumbsUp      6
dtype: int64


Now we will use Series.mean() function to find the mean of the given series object.

In [29]:

# return the mean 
result = sr.mean() 
  
# Print the result 
print(result) 


15.5


 Use Series.mean() function to find the mean of the underlying data in the given series object. The given series object also contains some missing values.

In [30]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the Series 
sr = pd.Series([19.5, 16.8, None, 22.78, 16.8, 20.124, None, 18.1002, 19.5]) 
  
# Print the series 
print(sr) 


0    19.5000
1    16.8000
2        NaN
3    22.7800
4    16.8000
5    20.1240
6        NaN
7    18.1002
8    19.5000
dtype: float64


In [31]:

# return the mean 
# skip all the missing values 
result = sr.mean(skipna = True) 
  
# Print the result 
print(result) 


19.086314285714284


#### Pandas dataframe.mad()

Pandas dataframe.mad() function return the mean absolute deviation of the values for the requested axis. The mean absolute deviation of a dataset is the average distance between each data point and the mean. It gives us an idea about the variability in a dataset.

    Syntax: DataFrame.mad(axis=None, skipna=None, level=None)


    Parameters :
    axis : {index (0), columns (1)}
    skipna : Exclude NA/null values when computing the result
    level : If the axis is a MultiIndex (hierarchical), count along a particular level, collapsing into a Series
    numeric_only : Include only float, int, boolean columns. If None, will attempt to use everything, then use only numeric data. Not implemented for Series.

    Returns : mad : Series or DataFrame (if level specified)


 Use mad() function to find the mean absolute deviation of the values over the index axis

In [32]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the dataframe  
df = pd.DataFrame({"A":[12, 4, 5, 44, 1], 
                   "B":[5, 2, 54, 3, 2],  
                   "C":[20, 16, 7, 3, 8], 
                   "D":[14, 3, 17, 2, 6]}) 
  
# Print the dataframe 
df 


A   B   C   D
0  12   5  20  14
1   4   2  16   3
2   5  54   7  17
3  44   3   3   2
4   1   2   8   6

In [33]:

# find the mean absolute deviation  
# over the index axis 
df.mad(axis = 0) 


A    12.32
B    16.32
C     5.76
D     5.68
dtype: float64

Use mad() function to find the mean absolute deviation of values over the column axis which is having some Na values in it.

In [34]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the dataframe  
df = pd.DataFrame({"A":[12, 4, 5, None, 1], 
                   "B":[7, 2, 54, 3, None], 
                   "C":[20, 16, 11, 3, 8],  
                   "D":[14, 3, None, 2, 6]}) 
  
# To find the mean absolute deviation 
# skip the Na values when finding the mad value 
df.mad(axis = 1, skipna = True) 


0     3.750000
1     4.875000
2    20.444444
3     0.444444
4     2.666667
dtype: float64

#### Pandas Series.mad() to calculate Mean Absolute Deviation of a Series

Pandas provide a method to make Calculation of MAD (Mean Absolute Deviation) very easy. MAD is defined as average distance between each value and mean.

 Syntax: Series.mad(axis=None, skipna=None, level=None)

Parameters:
axis: 0 or ‘index’ for row wise operation and 1 or ‘columns’ for column wise operation.
skipna: Includes NaN values too if False, Result will also be NaN even if a single Null value is included.
level: Defines level name or number in case of multilevel series.

Return Type: Float valu

In this example, a Series is created from a Python List using Pandas .Series() method. The .mad() method is called on series with all default parameters.

In [35]:

# importing pandas module  
import pandas as pd  
    
# importing numpy module  
import numpy as np  
    
# creating list 
list =[5, 12, 1, 0, 4, 22, 15, 3, 9] 
  
# creating series 
series = pd.Series(list) 
  
# calling .mad() method 
result = series.mad() 
  
# display 
result 


5.876543209876543

Explanation:

    Calculating Mean of series mean = (5+12+1+0+4+22+15+3+9) / 9 = 7.8888

    MAD = | (5-7.88)+(12-7.88)+(1-7.88)+(0-7.88)+(4-7.88)+(22-7.88)+(15-7.88)+(3-7.88)+(9-7.88)) | / 9.00

    MAD = (2.88 + 4.12 + 6.88 + 7.88 + 3.88 + 14.12 + 7.12 + 4.88 + 1.12) / 9.00

    MAD = 5.8755 (More accurately = 5.876543209876543)

#### Pandas dataframe.sem()

Pandas dataframe.sem() function return unbiased standard error of the mean over requested axis. The standard error (SE) of a statistic (usually an estimate of a parameter) is the standard deviation of its sampling distribution[1] or an estimate of that standard deviation. If the parameter or the statistic is the mean, it is called the standard error of the mean (SEM).

     Syntax : DataFrame.sem(axis=None, skipna=None, level=None, ddof=1, numeric_only=None, **kwargs)


    Parameters :
    axis : {index (0), columns (1)}
    skipna : Exclude NA/null values. If an entire row/column is NA, the result will be NA
    level : If the axis is a MultiIndex (hierarchical), count along a particular level, collapsing into a Series
    ddof : Delta Degrees of Freedom. The divisor used in calculations is N – ddof, where N represents the number of elements.
    numeric_only : Include only float, int, boolean columns. If None, will attempt to use everything, then use only numeric data. Not implemented for Series

    Return : sem : Series or DataFrame (if level specified)


 Use sem() function to find the standard error of the mean of the given dataframe over the index axis.

In [36]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the dataframe  
df = pd.read_csv("nba.csv") 
  
# Print the dataframe 
df 


Name                    Team  Number Position   Age  \
0              Avery Bradley          Boston Celtics     0.0       PG  25.0   
1                Jae Crowder          Boston Celtics    99.0       SF  25.0   
2               John Holland          Boston Celtics    30.0       SG  27.0   
3                R.J. Hunter          Boston Celtics    28.0       SG  22.0   
4              Jonas Jerebko          Boston Celtics     8.0       PF  29.0   
5               Amir Johnson          Boston Celtics    90.0       PF  29.0   
6              Jordan Mickey          Boston Celtics    55.0       PF  21.0   
7               Kelly Olynyk          Boston Celtics    41.0        C  25.0   
8               Terry Rozier          Boston Celtics    12.0       PG  22.0   
9               Marcus Smart          Boston Celtics    36.0       PG  22.0   
10           Jared Sullinger          Boston Celtics     7.0        C  24.0   
11             Isaiah Thomas          Boston Celtics     4.0       PG  27.0   
12               Evan Turner          Boston Celtics    11.0       SG  27.0   
13               James Young          Boston Celtics    13.0       SG  20.0   
14              Tyler Zeller          Boston Celtics    44.0        C  26.0   
15          Bojan Bogdanovic           Brooklyn Nets    44.0       SG  27.0   
16              Markel Brown           Brooklyn Nets    22.0       SG  24.0   
17           Wayne Ellington           Brooklyn Nets    21.0       SG  28.0   
18   Rondae Hollis-Jefferson           Brooklyn Nets    24.0       SG  21.0   
19              Jarrett Jack           Brooklyn Nets     2.0       PG  32.0   
20            Sergey Karasev           Brooklyn Nets    10.0       SG  22.0   
21           Sean Kilpatrick           Brooklyn Nets     6.0       SG  26.0   
22              Shane Larkin           Brooklyn Nets     0.0       PG  23.0   
23               Brook Lopez           Brooklyn Nets    11.0        C  28.0   
24          Chris McCullough           Brooklyn Nets     1.0       PF  21.0   
25               Willie Reed           Brooklyn Nets    33.0       PF  26.0   
26           Thomas Robinson           Brooklyn Nets    41.0       PF  25.0   
27                Henry Sims           Brooklyn Nets    14.0        C  26.0   
28              Donald Sloan           Brooklyn Nets    15.0       PG  28.0   
29            Thaddeus Young           Brooklyn Nets    30.0       PF  27.0   
..                       ...                     ...     ...      ...   ...   
428          Al-Farouq Aminu  Portland Trail Blazers     8.0       SF  25.0   
429          Pat Connaughton  Portland Trail Blazers     5.0       SG  23.0   
430             Allen Crabbe  Portland Trail Blazers    23.0       SG  24.0   
431                 Ed Davis  Portland Trail Blazers    17.0        C  27.0   
432         Maurice Harkless  Portland Trail Blazers     4.0       SF  23.0   
433         Gerald Henderson  Portland Trail Blazers     9.0       SG  28.0   
434              Chris Kaman  Portland Trail Blazers    35.0        C  34.0   
435           Meyers Leonard  Portland Trail Blazers    11.0       PF  24.0   
436           Damian Lillard  Portland Trail Blazers     0.0       PG  25.0   
437            C.J. McCollum  Portland Trail Blazers     3.0       SG  24.0   
438             Luis Montero  Portland Trail Blazers    44.0       SG  23.0   
439            Mason Plumlee  Portland Trail Blazers    24.0        C  26.0   
440            Brian Roberts  Portland Trail Blazers     2.0       PG  30.0   
441              Noah Vonleh  Portland Trail Blazers    21.0       PF  20.0   
442            Trevor Booker               Utah Jazz    33.0       PF  28.0   
443               Trey Burke               Utah Jazz     3.0       PG  23.0   
444               Alec Burks               Utah Jazz    10.0       SG  24.0   
445               Dante Exum               Utah Jazz    11.0       PG  20.0   
446           Derrick Favors               Utah Jazz    15.0       PF  24.

In [37]:

# find standard error of the mean of all the columns 
df.sem(axis = 0) 


Number         0.746862
Age            0.206011
Weight         1.233459
Salary    247611.576815
dtype: float64

Notice, all the non-numeric columns and values are automatically not included in the calculation of the dataframe. We did not have to specifically input the numeric columns for the calculation of the standard error of the mean.

Use sem() function to find the standard error of the mean over the column axis. Also do not skip the NaN values in the calculation of the dataframe.

In [38]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the dataframe  
df = pd.read_csv("nba.csv") 
  
# Calculate the standard error of  
# the mean of all the rows in dataframe 
df.sem(axis = 1, skipna = False) 


0      1.932567e+06
1      1.698999e+06
2               NaN
3      2.871404e+05
4      1.249978e+06
5      2.999970e+06
6      2.927141e+05
7      5.412647e+05
8      4.560713e+05
9      8.577368e+05
10     6.422908e+05
11     1.728199e+06
12     8.563560e+05
13     4.374393e+05
14     6.542168e+05
15     8.563536e+05
16     2.112451e+05
17     3.749793e+05
18     3.338479e+05
19     1.574981e+06
20     3.999400e+05
21     3.353287e+04
22     3.749835e+05
23     4.922224e+06
24     2.850415e+05
25     2.367958e+05
26     2.453118e+05
27     2.367950e+05
28     2.367983e+05
29     2.808966e+06
           ...     
428    2.010703e+06
429    1.562538e+05
430    2.367976e+05
431    1.745177e+06
432    7.234946e+05
433    1.499979e+06
434    1.253972e+06
435    7.689467e+05
436    1.059053e+06
437    6.312711e+05
438    1.312523e+05
439    3.538563e+05
440    7.137179e+05
441    6.594066e+05
442    1.193726e+06
443    6.645419e+05
444    2.365850e+06
445    9.444116e+05
446    2.999975e+06


#### Pandas Series.value_counts()

Pandas Series.value_counts() function return a Series containing counts of unique values. The resulting object will be in descending order so that the first element is the most frequently-occurring element. Excludes NA values by default

    Syntax: Series.value_counts(normalize=False, sort=True, ascending=False, bins=None, dropna=True)

    Parameter :
    normalize : If True then the object returned will contain the relative frequencies of the unique values.
    sort : Sort by values.
    ascending : Sort in ascending order.
    bins : Rather than count values, group them into half-open bins, a convenience for pd.cut, only works with numeric data.
    dropna : Don’t include counts of NaN.

    Returns : counts : Series


Use Series.value_counts() function to find the unique value counts of each element in the given Series object.

In [39]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the Series 
sr = pd.Series(['New York', 'Chicago', 'Toronto', 'Lisbon', 'Rio', 'Chicago', 'Lisbon']) 
  
# Print the series 
print(sr) 


0    New York
1     Chicago
2     Toronto
3      Lisbon
4         Rio
5     Chicago
6      Lisbon
dtype: object


In [40]:

# find the value counts 
sr.value_counts() 


Chicago     2
Lisbon      2
Rio         1
Toronto     1
New York    1
dtype: int64

Use Series.value_counts() function to find the unique value counts of each element in the given Series object.

In [41]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the Series 
sr = pd.Series([100, 214, 325, 88, None, 325, None, 325, 100]) 
  
# Print the series 
print(sr) 


0    100.0
1    214.0
2    325.0
3     88.0
4      NaN
5    325.0
6      NaN
7    325.0
8    100.0
dtype: float64


In [42]:

# find the value counts 
sr.value_counts() 


325.0    3
100.0    2
88.0     1
214.0    1
dtype: int64

#### Pandas Index.value_counts()

Pandas Index.value_counts() function returns object containing counts of unique values. The resulting object will be in descending order so that the first element is the most frequently-occurring element. Excludes NA values by default.

    Syntax: Index.value_counts(normalize=False, sort=True, ascending=False, bins=None, dropna=True)


    Parameters :
    normalize : If True then the object returned will contain the relative frequencies of the unique values.
    sort : Sort by values
    ascending : Sort in ascending order
    bins : Rather than count values, group them into half-open bins, a convenience for pd.cut, only works with numeric data
    dropna : Don’t include counts of NaN.

    Returns : counts : Series 

Use Index.value_counts() function to count the number of unique values in the given Index.

In [43]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the index 
idx = pd.Index(['Harry', 'Mike', 'Arther', 'Nick', 
                'Harry', 'Arther'], name ='Student') 
  
# Print the Index 
print(idx) 


Index(['Harry', 'Mike', 'Arther', 'Nick', 'Harry', 'Arther'], dtype='object', name='Student')


In [44]:

# find the count of unique values in the index 
idx.value_counts() 


Arther    2
Harry     2
Nick      1
Mike      1
Name: Student, dtype: int64

 Use Index.value_counts() function to find the count of all unique values in the given index.

In [45]:

# importing pandas as pd 
import pandas as pd 
  
# Creating the index 
idx = pd.Index([21, 10, 30, 40, 50, 10, 50]) 
  
# Print the Index 
print(idx) 


Int64Index([21, 10, 30, 40, 50, 10, 50], dtype='int64')


In [46]:

# for finding the count of all  
# unique values in the index. 
idx.value_counts() 


10    2
50    2
30    1
21    1
40    1
dtype: int64